Ignore GTK debug logs in the unit test unless they're enabled

vadz opened the pull request:
Don't output "*** GTK log message while running" messages for every
g_debug() call when we can be sure that the debug messages themselves
will not be displayed because G_MESSAGES_DEBUG is not set.

This results in much more reasonable output from the test suite.

See #17400.

vadz committed:
Ignore GTK debug logs in the unit test unless they're enabled

Don't output "*** GTK log message while running" messages for every
g_debug() call when we can be sure that the debug messages themselves
will not be displayed because G_MESSAGES_DEBUG is not set.

This results in much more reasonable output from the test suite.

See #17400.

ojwb commented on tests/test.cpp:
```diff
@@ -490,6 +490,15 @@ wxTestGLogHandler(const gchar* domain,
+    if ( s_ignoreDebug == -1 )
+    {
+        s_ignoreDebug = !wxGetEnv("G_MESSAGES_DEBUG", NULL);
+    }
```
FWIW, you should just be able to do:

```
static int s_ignoreDebug = !wxGetEnv("G_MESSAGES_DEBUG", NULL);
```

Such variables get "initialized the first time control passes through their declaration":

https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables

vadz replied:
Oops, yes, you're right, of course. I initially tested the value of the env var to check if `domain` argument was present in it (or if it was equal to "all"), but then decided it wasn't worth it and removed this code, but forgot to simplify it completely.

paulcor reviewed (commented):
I didn't know about `G_MESSAGES_DEBUG`. GLib doesn't cache that lookup, not sure it's worth bothering. What's the advantage of using `wxGetEnv()` here instead of just `getenv()`?

vadz commented:
> I didn't know about `G_MESSAGES_DEBUG`. 

There doesn't seem to be any better way to know if `g_log_default_handler()` will log the message or not. Or at least I couldn't find it.

> GLib doesn't cache that lookup, not sure it's worth bothering. What's the advantage of using `wxGetEnv()` here instead of just `getenv()`?

Mostly consistency with the other existing checks in this file. But also the fact that initially I used the value (see [my reply above](https://github.com/wxWidgets/wxWidgets/pull/1625#discussion_r341108456)). Not sure if I should restore this code or simplify the check even more.

vadz committed:
Make checks for G_MESSAGES_DEBUG more precise

Check that the domain is present in G_MESSAGES_DEBUG (or it's set to a
special "all" value) instead of assuming that all messages will be
logged just because this environment variable is set.

vadz committed:
Remove extra new lines after GTK log messages

This seems unnecessary as the default log handler already outputs a new
line at the end of the message anyhow and at least some messages (e.g.
debug ones in GdkPixbuf) also contain an extra new line in them, so
adding another one here resulted in having at least one and sometimes
two extra blank lines.

vadz commented:
Here is the full version of the code (with one extra small fix for superfluous newlines).

Which one do you think is better?

paulcor commented:
I'm sure it works, but it seems overly complicated. If it were me, I'd just do this:
```diff
@@ -490,12 +490,16 @@ wxTestGLogHandler(const gchar* domain,
                   const gchar* message,
                   gpointer data)
 {
+    if (level == G_LOG_LEVEL_DEBUG)
+    {
+        static const char* s = getenv("G_MESSAGES_DEBUG");
+        if (s == NULL || (strcmp(s, "all") != 0 && strstr(s, domain) == NULL))
+            return;
+    }
     fprintf(stderr, "\n*** GTK log message while running %s(): ",
             wxGetCurrentTestName().c_str());
 
     g_log_default_handler(domain, level, message, data);
-
-    fprintf(stderr, "\n");
 }
 
 #endif // __WXGTK__
```

As an aside, in general I would avoid using things (like `wxString`) in the test machinery which are actually tested by the tests. Looks like that ship has sailed though...

vadz commented:
Sure, let's simplify it like this, I also balked at doing it like this, hence the initial simple -- but maybe too simple -- version. Thanks!

As for using `wxString`, I don't think it's really a concern. We suppose that basic stuff does work and if it really doesn't, better to know about it sooner rather than later.
